City-of-Helsinki/berth-reservations
# VEN-1351 | Application priorities

## Description :sparkles:
* Add an `ApplicationPriority` field to the `BaseApplication` model
* Sort the application first by `priority`, then by `created_at`
* When the user extends the validity of the application `ExtendBerthApplicationMutation`, set the priority to `LOW`
* Add a cronjob to reset the applications with LOW-priority every 1st of January

## Issues :bug:
### Closes :no_good_woman:
**[VEN-1351](https://helsinkisolutionoffice.atlassian.net/browse/VEN-1351):** Add ApplicationPriority to applications

## Testing :alembic:
### Automated tests :gear:️
```shell
$ pytest applications/tests/test_applications_berth_application_queries.py::test_berth_application_priority
$ pytest applications/tests/test_applications_mutations.py::test_extend_berth_application
$ pytest applications/tests/test_application_models.py::test_reset_application_priorities
```

### Manual testing :construction_worker_man:
1. Update an application to change the priority
```graphql
mutation Update($input: UpdateBerthApplicationInput!) {
  updateBerthApplication(input: $input) {
    berthApplication {
      createdAt
      priority
    }
  }
}
```
```json
{
  "input": {
    "id": "",
    "priority": "LOW"
  }
}
```
2. Check the application list to see that the priority takes effect
```graphql
query Applications {
  berthApplications {
    edges {
      node {
        id
        createdAt
        priority
      }
    }
  }
}
```

## Screenshots :camera_flash:
<img width="517" alt="image" src="https://user-images.githubusercontent.com/15201480/123167817-b1bcd180-d43c-11eb-9983-9c772755bc77.png">
